src/vidpf.rs

```diff
@@ -366,8 +381,48 @@ impl VidpfKey {
     }
 }
 
+impl Decode for VidpfKey {
+    fn decode(bytes: &mut Cursor<&[u8]>) -> Result<Self, CodecError> {
+        let id = u8::decode(bytes)?;
+        let mut value = [0; 16];
+        bytes.read_exact(&mut value)?;
+        Ok(VidpfKey {
+            id: match id {
+                0 => VidpfServerId::S0,
+                1 => VidpfServerId::S1,
+                _ => return Err(CodecError::UnexpectedValue),
+            },
+            value,
+        })
+    }
+}
+
+impl Encode for VidpfKey {
+    fn encode(&self, bytes: &mut Vec<u8>) -> Result<(), CodecError> {
+        bytes.push(match self.id {
+            VidpfServerId::S0 => 0u8,
+            VidpfServerId::S1 => 1u8,
+        });
+        bytes.extend_from_slice(&self.value[..]);
+        Ok(())
+    }
+
+    fn encoded_len(&self) -> Option<usize> {
+        Some(17)
+    }
+}
```

Comment:
It should always be the case that whoever is decoding a `VidpfKey` knows its role (i.e., the value of `VidpfServerId`), so I don't think this is needed. We should remove it because:
1. We can save a byte on the wire for each input share (this adds up)
2. We don't have to handle the case where the server tries to parse the wrong role

What we could do instead is add the `VidpfServerId` to the decoding parameter.